### ISSUE TITLE
Provider-specific code in base.py

### ISSUE THREAD

**Issue by themiurgo** [DenisCarriere/geocoder#78]
base.py contains provider specific code (Google Exception, MaxMind exception, ...). Any particular reasons for that? If not, it would be good to move that to the provider modules/classes.

**Comment by DenisCarriere** [comment]
Ok Antonio, I've finally got around to solve the ugly code in the `base.py` a few things will still be broken, but I've focused on the Google one and I'm working my way down the list.

Let me know if there is any other parts of the code that don't look "Elegant" Python language.

All those ugly lines of code got replaced by:
```python
    def tree(self): return defaultdict(self.tree)

    def _build_tree(self, content):
        if content:
            for key, value in content.items():
                self.parse[key] = value
```
And the exceptions are within the Provider code now, which will make sense when there's some provider issues with their JSON objects.

Thanks for this issue post, it took me a while to get around to it but I've finally found some time to tackle it.
;)
